File: v19_facade/java/Movie2.java
package v19_facade.java;

class Projector {
    public void on() {
        System.out.println("Turning on the projector...");
    }

    public void off() {
        System.out.println("Turning off the projector...");
    }
}

class DVDPlayer {
    public void play() {
        System.out.println("Playing the movie...");
    }

    public void stop() {
        System.out.println("Stopping the movie...");
    }
}

class SoundSystem {
    public void on() {
        System.out.println("Turning on the sound system...");
    }

    public void off() {
        System.out.println("Turning off the sound system...");
    }
}

class HomeTheaterFacade {
    private Projector projector;
    private DVDPlayer dvdPlayer;
    private SoundSystem soundSystem;

    public HomeTheaterFacade(Projector projector, DVDPlayer dvdPlayer, SoundSystem soundSystem) {
        this.projector = projector;
        this.dvdPlayer = dvdPlayer;
        this.soundSystem = soundSystem;
    }

    public void watchMovie() {
        projector.on();
        dvdPlayer.play();
        soundSystem.on();
    }

    public void stopMovie() {
        projector.off();
        dvdPlayer.stop();
        soundSystem.off();
    }
}

public class Movie2 {
    public static void main(String[] args) {
        Projector projector = new Projector();
        DVDPlayer dvdPlayer = new DVDPlayer();
        SoundSystem soundSystem = new SoundSystem();

        HomeTheaterFacade homeTheater = new HomeTheaterFacade(projector, dvdPlayer, soundSystem);
        homeTheater.watchMovie();
    }
}

File: v19_facade/python/movie2.py
class Projector:
    def on(self):
        print("Turning on the projector...")

    def off(self):
        print("Turning off the projector...")

class DVDPlayer:
    def play(self):
        print("Playing the movie...")

    def stop(self):
        print("Stopping the movie...")

class SoundSystem:
    def on(self):
        print("Turning on the sound system...")

    def off(self):
        print("Turning off the sound system...")

class HomeTheaterFacade:
    def __init__(self, projector, dvd_player, sound_system):
        self.projector = projector
        self.dvd_player = dvd_player
        self.sound_system = sound_system

    def watch_movie(self):
        self.projector.on()
        self.dvd_player.play()
        self.sound_system.on()

    def stop_movie(self):
        self.projector.off()
        self.dvd_player.stop()
        self.sound_system.off()

# Client code
projector = Projector()
dvd_player = DVDPlayer()
sound_system = SoundSystem()

home_theater = HomeTheaterFacade(projector, dvd_player, sound_system)
home_theater.watch_movie()
